### PR TITLE
Improve Jax INLA numerical stability

### PR DESCRIPTION
**Pull request by constinit**
And add test to ensure it is stable.

- We use the eigen factorization to compute the gradient during the
Newton steps
- We slightly modify the jax_faster routines for improved stability
- We stabilize the final log probability sum by substracting the smallest
element first

### CODE REVIEW

File: research/berry/berrylib/test_berry.py [hunk 1]
```diff
@@ -182,7 +187,8 @@ def test_fast_inla(method, N=10, iterations=1):
         start = time.time()
         out = inla_model.inference(y_i, n_i, method=method)
         end = time.time()
-        np.testing.assert_allclose(out[0][0].sum(), 2111.808, rtol=1e-3)
```
Comment (tbenthompson) [inline research/berry/berrylib/test_berry.py:185]: Wait, this line was failing before and you just deleted it? Was this just a wrong check?

File: research/berry/berrylib/test_berry.py [hunk 2]
```diff
@@ -226,6 +229,24 @@ def logistic(x):
     )
 
 
+def test_fast_inla_same_results(N=1, iterations=1000):
```
Comment (tbenthompson) [inline research/berry/berrylib/test_berry.py:232]: Thanks, this seems like a nice check!

**Review by tbenthompson — APPROVED** [review]
Nice. Just one question on the test. Otherwise, merge any time!

**Review by tbenthompson — CHANGES_REQUESTED** [review]
I investigated the `test_fast_inla` JAX problems a bit more. I think we need to make some changes here:

The `logistic(...)` calls in `test_fast_inla` are making the test useless. I don't understand what motivated this change, but it definitely needs to be reversed. The removal of the check against `2111.808` should also be reversed. See below for values where this is a problem. It's masking the problems with small values of sigma2.

You're fighting windmills here as far as numerical accuracy for these small values of sigma2. There just aren't enough digits of precision in a 32 bit float to capture the covariance matrix when we're adding 1e-6 to the diagonal of a matrix full of 100s. Basically we're considering a degenerate case where sigma2 is effectively zero. I think there are likely to be two viable solutions here:
1. Make some kind of analytic simplification of the linear algebra operations involving the covariance and precision matrices for these degenerate cases.
2. For small sigma2, use the single trial "approximation". This should be almost exact. Basically, if sigma2 < 1e-4 or so, we sum y and sum n and then treat the problem like it only has one arm! See the example of this approximation at the end of this comment. **this is the much better solution!**

**logistic problems**

If I look at the output for sigma2_post, the differences with the correct values are quite stark, but only for the small values of sigma2.
```
sigma2_post[0]
DeviceArray([1.54817352e+02, 4.83758362e+02, 8.58758362e+02,
             5.08642273e+02, 1.30386200e+02, 2.20411873e+01,
             3.15194583e+00, 5.50986826e-01, 2.68367738e-01,
             1.23575605e-01, 1.13334851e-02, 5.94810990e-04,
             4.01075886e-05, 4.92795425e-06, 1.41589533e-06],            dtype=float32)
correct
array([1.25954474e+02, 4.52520893e+02, 8.66625278e+02, 5.08333300e+02,
       1.30365045e+02, 2.20403048e+01, 3.15183578e+00, 5.50967224e-01,
       3.68365061e-01, 1.23585852e-01, 1.13330444e-02, 5.94800210e-04,
       4.01075571e-05, 4.92782335e-06, 1.41605356e-06])
```

If I call logistic:
```
logistic(sigma2_post[0])
DeviceArray([1.        , 1.        , 1.        , 1.        , 1.        ,
             1.        , 0.9589853 , 0.63436455, 0.5666922 , 0.53085464,
             0.50283337, 0.5001487 , 0.50001   , 0.50000125, 0.50000036],            dtype=float32)
```
This means that the logistic is saturating for the first few indices of the sigma2 posterior and making the check useless.

**single trial degeneracy**

```
import berrylib.fast_inla as fast_inla
import numpy as np
fi = fast_inla.FastINLA(n_arms=3, sigma2_n=1, sigma2_bounds=[1e-6, 1e-5])
fi.sigma2_rule
y = np.array([[3,5,7]])
n = np.array([[35,35,35]])
print('separately')
print(fi.numpy_inference(y, n))

y_degen = y.sum(axis=1)[:,None]
n_degen = n.sum(axis=1)[:, None]
fi_degen = fast_inla.FastINLA(n_arms=1, sigma2_n=1, sigma2_bounds=[1e-6, 1e-5])
print('together')
print(fi_degen.numpy_inference(y_degen, n_degen))
```

output. see all the values are the same!
```
separately
(array([[137335.97380571]]), array([[0.92691033, 0.92691348, 0.92691663]]), array([[[-0.94477537, -0.94476904, -0.94476272]]]), array([[[0.27880711, 0.27880711, 0.27880711]]]), array([[[[-0.07773341, -0.07773024, -0.07773024],
         [-0.07773024, -0.07773341, -0.07773024],
         [-0.07773024, -0.07773024, -0.07773341]]]]))
together
(array([[137335.97380571]]), array([[0.92691622]]), array([[[-0.94476904]]]), array([[[0.27880333]]]), array([[[[-0.0777313]]]]))
```

File: research/berry/berrylib/test_berry.py [hunk 3]
```diff
@@ -192,9 +198,6 @@ def test_fast_inla(method, N=10, iterations=1):
 
     sigma2_post, exceedances, theta_max, theta_sigma = out
 
-    def logistic(x):
-        return jax.scipy.special.expit(x)
-
     # Do our comparisons in probability space.
     np.testing.assert_allclose(
         logistic(theta_max[0, 12]),
```
Comment (tbenthompson) [inline research/berry/berrylib/test_berry.py:203]: I think the logistic transform is fine here for `theta_max`. It's only a problem for `sigma2_post`. Those values are not in logit space. Those are densities for the sigma2 posterior so it doesn't make sense to logistic them.

**Comment by tbenthompson** [comment]
I thought it might be useful to demonstrate how we can accurately find `theta_max` here without ever using the value of sigma2. This snippet is a follow on to the one above:
```
import scipy.optimize
import scipy.special
mu_0 = -1.34
mu_sig2 = 100
def obj(theta):
    theta_adj = theta + fi_degen.logit_p1
    exp_theta_adj = np.exp(theta_adj)
    y = y_degen[0,0]
    n = n_degen[0,0]
    return -(
        0.5 * (theta - mu_0) ** 2 / -mu_sig2
        + theta_adj * y - n * np.log(exp_theta_adj + 1)
    )
optimum = scipy.optimize.minimize(obj, -0.5, tol=1e-14)['x']
print('theta_max', optimum[0])
print('error', (optimum - fi_degen.numpy_inference(y_degen, n_degen)[2])[0,0,0])
```

We get the same `theta_max` to almost 9 digits. 
```
theta_max -0.9447690480267578
error -3.5866418723884408e-09
```
